File: impl/golang/internal/registry/registration.go
package registry

import (
	"context"
	"math/big"

	"github.com/ethereum/go-ethereum/common"
	"github.com/neuron-sdk/neuron-go-sdk/internal/keylib"
)

// Registration represents a Child's registration in an EIP-8004 registry,
// linked to an extended NFT (ERC-721). One registration per (childAddress,
// registryAddress) per FR-R05.
type Registration struct {
	registryAddress keylib.EVMAddress
	childAddress    keylib.EVMAddress
	tokenId         *big.Int
	agentURI        AgentURI
	chainId         uint64
}

// RegistryAddress returns the registry's EVM address.
func (r *Registration) RegistryAddress() keylib.EVMAddress { return r.registryAddress }

// ChildAddress returns the child's EVM address.
func (r *Registration) ChildAddress() keylib.EVMAddress { return r.childAddress }

// TokenId returns a defensive copy of the registration token ID.
func (r *Registration) TokenId() *big.Int {
	if r.tokenId == nil {
		return nil
	}
	return new(big.Int).Set(r.tokenId)
}

// AgentURI returns the agent URI for this registration.
func (r *Registration) AgentURI() AgentURI { return r.agentURI }

// ChainId returns the chain ID of the registry.
func (r *Registration) ChainId() uint64 { return r.chainId }

// RegistrationResult is the result of a registration operation (create,
// update, or revoke).
type RegistrationResult struct {
	tokenId         *big.Int
	transactionHash string
	childAddress    keylib.EVMAddress
	registryAddress keylib.EVMAddress
	chainId         uint64
	agentURIString  string
}

// TokenId returns a defensive copy of the registration token ID.
func (r *RegistrationResult) TokenId() *big.Int {
	if r.tokenId == nil {
		return nil
	}
	return new(big.Int).Set(r.tokenId)
}

// TransactionHash returns the transaction hash.
func (r *RegistrationResult) TransactionHash() string { return r.transactionHash }

// ChildAddress returns the child's EVM address.
func (r *RegistrationResult) ChildAddress() keylib.EVMAddress { return r.childAddress }

// RegistryAddress returns the registry's EVM address.
func (r *RegistrationResult) RegistryAddress() keylib.EVMAddress { return r.registryAddress }

// ChainId returns the chain ID of the registry.
func (r *RegistrationResult) ChainId() uint64 { return r.chainId }

// AgentURIString returns the serialized agent URI JSON string.
func (r *RegistrationResult) AgentURIString() string { return r.agentURIString }

// Register mints a new registration NFT for a Child agent in the given
// EIP-8004 Identity Registry.
//
// Steps:
//  1. Derive childAddress from childKey.PublicKey().EVMAddress()
//  2. Validate agentURI via ValidateRegistrationCompleteness
//  3. Check for duplicate registration (tokenOfOwnerByIndex returns existing token)
//  4. Serialize agentURI to JSON
//  5. Call contract.Register()
//  6. Verify ownerOf matches childAddress
//  7. Return RegistrationResult
func Register(
	ctx context.Context,
	childKey *keylib.NeuronPrivateKey,
	registryAddr keylib.EVMAddress,
	chainId uint64,
	agentURI AgentURI,
	contract RegistryContract,
	admission AdmissionChecker,
	parentDID string,
) (RegistrationResult, error) {
	childPub := childKey.PublicKey()
	childAddress := childPub.EVMAddress()
	childCommon := common.BytesToAddress(childAddress.Bytes())

	// Validate agentURI completeness.
	valid, validationErrors := ValidateRegistrationCompleteness(agentURI, childPub)
	if !valid {
		return RegistrationResult{}, IncompleteRegistration{
			Detail: validationErrors[0].Error(),
		}
	}

	// Check admission policy.
	if admitted, err := admission.IsAdmitted(childAddress.Hex(), parentDID); !admitted {
		if err != nil {
			return RegistrationResult{}, err
		}
		return RegistrationResult{}, UnauthorizedOperation{
			CallerRole: "parent",
			Operation:  "register",
		}
	}

	// Check for duplicate registration.
	_, err := contract.TokenOfOwnerByIndex(ctx, childCommon, big.NewInt(0))
	if err == nil {
		// Token exists — duplicate registration.
		return RegistrationResult{}, DuplicateRegistration{
			ChildAddress:    childAddress.Hex(),
			RegistryAddress: registryAddr.Hex(),
		}
	}

	// Serialize agentURI to JSON.
	agentURIJSON, err := agentURI.ToJSON()
	if err != nil {
		return RegistrationResult{}, RegistryUnavailable{Detail: "failed to serialize agentURI: " + err.Error()}
	}

	// Call contract.Register().
	tokenId, txHash, err := contract.Register(ctx, childCommon, agentURIJSON)
	if err != nil {
		return RegistrationResult{}, RegistryUnavailable{Detail: err.Error()}
	}

	// Verify ownerOf matches childAddress.
	owner, err := contract.OwnerOf(ctx, tokenId)
	if err != nil {
		return RegistrationResult{}, RegistryUnavailable{Detail: "failed to verify ownership: " + err.Error()}
	}
	if owner != childCommon {
		return RegistrationResult{}, ProofOfControlFailed{
			Expected: childAddress.Hex(),
			Actual:   owner.Hex(),
		}
	}

	return RegistrationResult{
		tokenId:         tokenId,
		transactionHash: txHash,
		childAddress:    childAddress,
		registryAddress: registryAddr,
		chainId:         chainId,
		agentURIString:  agentURIJSON,
	}, nil
}

// UpdateRegistration updates the agentURI for an existing registration.
//
// Steps:
//  1. Derive caller address
//  2. Verify caller is owner or approved operator
//  3. Validate newAgentURI
//  4. Call contract.UpdateAgentURI()
//  5. Return RegistrationResult
func UpdateRegistration(
	ctx context.Context,
	childKey *keylib.NeuronPrivateKey,
	registryAddr keylib.EVMAddress,
	chainId uint64,
	tokenId *big.Int,
	newAgentURI AgentURI,
	contract RegistryContract,
) (RegistrationResult, error) {
	childPub := childKey.PublicKey()
	childAddress := childPub.EVMAddress()
	callerCommon := common.BytesToAddress(childAddress.Bytes())

	// Verify caller is owner or approved operator.
	owner, err := contract.OwnerOf(ctx, tokenId)
	if err != nil {
		return RegistrationResult{}, RegistrationNotFound{
			Detail: "token not found: " + err.Error(),
		}
	}

	if owner != callerCommon {
		// Check if caller is an approved operator.
		approved, err := contract.GetApproved(ctx, tokenId)
		if err != nil || approved != callerCommon {
			approvedForAll, err := contract.IsApprovedForAll(ctx, owner, callerCommon)
			if err != nil || !approvedForAll {
				return RegistrationResult{}, UnauthorizedOperation{
					CallerRole: "caller",
					Operation:  "updateAgentURI",
				}
			}
		}
	}

	// Validate newAgentURI completeness.
	valid, validationErrors := ValidateRegistrationCompleteness(newAgentURI, childPub)
	if !valid {
		return RegistrationResult{}, IncompleteRegistration{
			Detail: validationErrors[0].Error(),
		}
	}

	// Serialize and update.
	agentURIJSON, err := newAgentURI.ToJSON()
	if err != nil {
		return RegistrationResult{}, RegistryUnavailable{Detail: "failed to serialize agentURI: " + err.Error()}
	}

	txHash, err := contract.UpdateAgentURI(ctx, callerCommon, tokenId, agentURIJSON)
	if err != nil {
		return RegistrationResult{}, RegistryUnavailable{Detail: err.Error()}
	}

	return RegistrationResult{
		tokenId:         tokenId,
		transactionHash: txHash,
		childAddress:    childAddress,
		registryAddress: registryAddr,
		chainId:         chainId,
		agentURIString:  agentURIJSON,
	}, nil
}

// RevokeRegistration burns (destroys) a registration NFT, revoking the
// Child's registration in the registry.
//
// Steps:
//  1. Derive caller address
//  2. Verify caller is owner
//  3. Call contract.Burn()
//  4. Return transactionHash
func RevokeRegistration(
	ctx context.Context,
	childKey *keylib.NeuronPrivateKey,
	registryAddr keylib.EVMAddress,
	chainId uint64,
	tokenId *big.Int,
	contract RegistryContract,
) (string, error) {
	childAddress := childKey.PublicKey().EVMAddress()
	callerCommon := common.BytesToAddress(childAddress.Bytes())

	// Verify caller is owner.
	owner, err := contract.OwnerOf(ctx, tokenId)
	if err != nil {
		return "", RegistrationNotFound{
			Detail: "token not found: " + err.Error(),
		}
	}

	if owner != callerCommon {
		return "", UnauthorizedOperation{
			CallerRole: "caller",
			Operation:  "burn",
		}
	}

	txHash, err := contract.Burn(ctx, callerCommon, tokenId)
	if err != nil {
		return "", RegistryUnavailable{Detail: err.Error()}
	}

	return txHash, nil
}

File: impl/typescript/src/registry/registration.ts
/**
 * Registration operations -- register, update, revoke, lookup.
 *
 * Spec reference: 003 spec.md
 *   - FR-R01: Registry is EIP-8004 smart contract.
 *   - FR-R06: register() signed by Child's NeuronPrivateKey (proof-of-control).
 *   - FR-R07: Lookup failure is a documented error.
 *   - FR-R08: AgentURI validated for completeness before register/update.
 *   - FR-R10: NFT owner is Child's EVMAddress after mint.
 *   - FR-R11: Role boundaries for operations.
 *
 * Spec reference: 003-to-007 mapping
 *   - register() -> 007 register()
 *   - updateRegistration() -> 007 updateAgentURI()
 *   - revokeRegistration() -> 007 revoke()
 *   - lookupRegistration() -> 007 lookup()
 */

import type { NeuronPrivateKey } from '../keylib/private-key.js';
import type { RegistryContract } from './contract.js';
import type { AgentURI, RegistrationResult, Registration, LookupKey } from './types.js';
import { validateRegistrationCompleteness } from './validation.js';
import { serializeAgentURI } from './agent-uri.js';
import {
  registrationFailed,
  updateFailed,
  revocationFailed,
  lookupFailed,
  invalidAgentURI,
} from './errors.js';
import { RegistryError } from './errors.js';

/**
 * Register a Child agent in an EIP-8004 registry.
 *
 * FR-R06: The Child's NeuronPrivateKey provides proof-of-control.
 * FR-R08: Validates AgentURI completeness before submitting.
 * FR-R10: The resulting NFT is owned by the Child's EVMAddress.
 *
 * Workflow:
 * 1. Validate AgentURI (V-REG-01..07, V-REG-11, V-REG-12)
 * 2. Serialize AgentURI to JSON
 * 3. Call contract.register() (on-chain transaction)
 *
 * @param childKey - The Child's NeuronPrivateKey (for proof-of-control and validation)
 * @param contract - RegistryContract adapter for the target chain
 * @param agentURI - The AgentURI containing all required services
 * @returns RegistrationResult with tokenId, transactionHash, and metadata
 * @throws RegistryError NEURON-REG-005 if AgentURI validation fails
 * @throws RegistryError NEURON-REG-001 if on-chain registration fails
 */
export async function register(
  childKey: NeuronPrivateKey,
  contract: RegistryContract,
  agentURI: AgentURI,
): Promise<RegistrationResult> {
  // FR-R08: Validate completeness before submission
  const childPubKey = childKey.publicKey();
  const validation = validateRegistrationCompleteness(agentURI, childPubKey);

  if (!validation.valid) {
    const messages = validation.errors.map(e => `[${e.rule}] ${e.message}`);
    throw invalidAgentURI(
      `AgentURI validation failed: ${messages.join('; ')}`,
    );
  }

  // Serialize and submit
  const uriJson = serializeAgentURI(agentURI);

  try {
    return await contract.register(uriJson);
  } catch (e) {
    if (e instanceof RegistryError) {
      throw e;
    }
    const cause = e instanceof Error ? e : new Error(String(e));
    throw registrationFailed('On-chain register() call failed', cause);
  }
}

/**
 * Update the AgentURI of an existing registration.
 *
 * FR-R06: Post-mint, updateAgentURI is governed by ERC-721 token ownership.
 * FR-R08: Validates the new AgentURI for completeness.
 * FR-R11b: Only the registered agent or approved operator may update.
 *
 * @param callerKey - The caller's NeuronPrivateKey (Child or delegated operator)
 * @param contract - RegistryContract adapter for the target chain
 * @param tokenId - The ERC-721 token ID of the registration to update
 * @param newAgentURI - The new AgentURI containing all required services
 * @returns RegistrationResult with the updated metadata
 * @throws RegistryError NEURON-REG-005 if AgentURI validation fails
 * @throws RegistryError NEURON-REG-003 if on-chain update fails
 * @throws RegistryError NEURON-REG-006 if caller is not authorized
 */
export async function updateRegistration(
  callerKey: NeuronPrivateKey,
  contract: RegistryContract,
  tokenId: bigint,
  newAgentURI: AgentURI,
): Promise<RegistrationResult> {
  // FR-R08: Validate completeness before submission
  const callerPubKey = callerKey.publicKey();
  const validation = validateRegistrationCompleteness(newAgentURI, callerPubKey);

  if (!validation.valid) {
    const messages = validation.errors.map(e => `[${e.rule}] ${e.message}`);
    throw invalidAgentURI(
      `AgentURI validation failed: ${messages.join('; ')}`,
    );
  }

  const uriJson = serializeAgentURI(newAgentURI);

  try {
    return await contract.updateAgentURI(tokenId, uriJson);
  } catch (e) {
    if (e instanceof RegistryError) {
      throw e;
    }
    const cause = e instanceof Error ? e : new Error(String(e));
    throw updateFailed('On-chain updateAgentURI() call failed', cause);
  }
}

/**
 * Revoke (burn) a registration in the registry.
 *
 * FR-R11b: Only the registered agent may revoke its own registration.
 *
 * @param childKey - The Child's NeuronPrivateKey
 * @param contract - RegistryContract adapter for the target chain
 * @param tokenId - The ERC-721 token ID of the registration to revoke
 * @returns The on-chain transaction hash
 * @throws RegistryError NEURON-REG-004 if on-chain revocation fails
 * @throws RegistryError NEURON-REG-006 if caller is not authorized
 */
export async function revokeRegistration(
  _childKey: NeuronPrivateKey,
  contract: RegistryContract,
  tokenId: bigint,
): Promise<string> {
  try {
    return await contract.revoke(tokenId);
  } catch (e) {
    if (e instanceof RegistryError) {
      throw e;
    }
    const cause = e instanceof Error ? e : new Error(String(e));
    throw revocationFailed('On-chain revoke() call failed', cause);
  }
}

/**
 * Look up a registration in the registry.
 *
 * FR-R04: Lookup by (registry + Child EVM address) or (registry + external id).
 * FR-R07: Returns null when not found; throws on network errors.
 *
 * @param contract - RegistryContract adapter for the target chain
 * @param key - Lookup discriminant (byAddress or byExternalId)
 * @returns Registration if found, null if not found
 * @throws RegistryError NEURON-REG-002 if lookup fails (network error, timeout)
 */
export async function lookupRegistration(
  contract: RegistryContract,
  key: LookupKey,
): Promise<Registration | null> {
  try {
    return await contract.lookup(key);
  } catch (e) {
    if (e instanceof RegistryError) {
      throw e;
    }
    const cause = e instanceof Error ? e : new Error(String(e));
    throw lookupFailed('Registry lookup failed', cause);
  }
}
